spec/Classes.tex

```diff
@@ -271,6 +267,72 @@ \subsection{Class Types}
 \end{chapelexample}
 
 
+\subsection{Nilable Classes}
+\label{Nilable_Classes}
+
+Variables of class type cannot store \chpl{nil} unless the class type is
+nilable. To declare a nilable class type, use the \chpl{?} operator to
+create a nilable type. For example, if \chpl{C} is a class type, then
+\chpl{C?} indicates nilable the class type with generic management. The
+\chpl{?} operator can be combined with memory management specifiers as
+well. For example, \chpl{borrowed C?} indicates a nilable class using the
+\chpl{borrowed} memory management strategy. Note that the \chpl{?}
+operator only applies to types.
+
+The postfix \chpl{!} operator applies to a type or a value. When applied
+to a type, it returns the non-nilable borrowed type. When applied to a
+value, it asserts that the value is not \chpl{nil} and returns that value
+as a non-nilable borrowed type. If the value was in fact \chpl{nil}, it
+may halt.
+
+An alternative to \chpl{!} is to use a cast to a non-nilable type. Such a
+cast will throw \chpl{NilClassError} if the value was in fact \chpl{nil}.
+See~\rsec{Explicit_Class_Conversions}.
+
+Non-nilable class types are implicitly convertible to nilable class
+types. See~\rsec{Implicit_Class_Conversions}.
+
+Class methods generally expect a receiver of type \chpl{borrowed C}
+(see~\rsec{Class_Methods}).  Since such a class method call might involve
+dynamic dispatch, it is a program error to call a class method on a class
+receiver storing \chpl{nil}. The language helps to identify this error in
+two different ways, depending on whether or not the module is a
+\chpl{prototype} module (recall that implicit modules are
+\chpl{prototype} modules).
+
+For modules that are not prototype modules, the compiler will not
+resolve calls to such class methods if the receiver has nilable type.
+If the programmer knows that the receiver cannot store \chpl{nil} at that
+moment, they can use \chpl{!} to assert that the reciever is not
+\chpl{nil} and to convert it to the non-nilable borrowed type. For
+example:
+
+\begin{chapelexample}{nilable-classes-bang.chpl}
+\begin{chapel}
+class C {
+  proc C.method() { }
+}
+var c: owned C? = makeMyClass();
+
+// The following call is allowed only in prototype modules,
+// in which case it is equivalent to c!.method()
+c.method();
+
+// This pattern is appropriate in libraries
+if c != nil {
+  c!.method(); // c! converts from 'owned C?' to 'borrowed C'
+}
+\end{chapel}
+\begin{chapeloutput}
+\end{chapeloutput}
+\end{chapelexample}
+
+\label{Methods_On_Nilable_In_Prototype_Modules}
+
+Within a \chpl{prototype} module, the compiler will implicitly convert
+a nilable method receiver to the non-nilable type by adding a \chpl{!}
+call.
```

Comment:
I think the `borrow` method is an exception. It is applicable to `nil` and if so produces `nil`.

If this is the case, please explain it in this subsection.